### PR TITLE
added max-of-n

### PR DESCRIPTION
**Pull request by euanong**

### CODE REVIEW

File: gbmi/exp_max_of_n/train.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        return model
+
+    @staticmethod
+    def loss_fn(
+        logits: Float[Tensor, "batch pos d_vocab"],
+        tokens: Integer[Tensor, "batch pos"],
+    ) -> Float[Tensor, ""]:
+        logits = logits[:, -1, :]
+        true_maximum = torch.max(tokens, dim=1)[0]
+        log_probs = logits.log_softmax(-1)
```
Comment (JasonGross) [inline gbmi/exp_max_of_n/train.py:93]: Do we want to do the more accurate (currently untested)
```suggestion
        # log_softmax is only around to around 2e-7, cf https://github.com/pytorch/pytorch/issues/113708
        # we can get better precision by using log1p
        logits_max_idxs = logits.argmax(dim=-1, keepdim=True)
        logits_centered = logits - logits.gather(dim=-1, index=logits_max_idxs)
        logits_exp = logits_centered.exp()
        # logits_exp[max] will be 1, so we can zero it and use log1p(x) = log(1 + x)
        logits_exp.scatter_(dim=-1, index=logits_max_idxs, 0)
        log_probs = logits_centered - logits_exp.sum(dim=-1, keepdim=True).log1p()
```

File: gbmi/exp_max_of_n/train.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        return self.run_batch(batch, prefix="")
+
+    def test_step(self, batch, batch_idx):
+        self.run_batch(batch, prefix="test_")
+
+    def configure_optimizers(self):
+        return torch.optim.Adam(
+            self.parameters(),
+            lr=self.config.optimizer_kwargs["lr"],
+            betas=self.config.optimizer_kwargs["betas"],
```
Comment (JasonGross) [inline gbmi/exp_max_of_n/train.py:128]: More compactly generalizable alternative (pick whichever seems better to you):
```suggestion
            **{k:self.config.optimizer_kwargs[k] for k in ("lr", "betas")}
```

File: gbmi/exp_max_of_n/train.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+class MaxOfNDataModule(DataModule):
+    data_train: Dataset[Integer[Tensor, "seq_len"]]
+    data_test: Dataset[Integer[Tensor, "seq_len"]]
+    batch_size: Optional[int]
+
+    def __init__(self, config: Config[MaxOfN]):
+        super().__init__(config)
+        self.config = config
+        self.seq_len = config.n_ctx
+        self.dataset_seed = config.seed * 10 + 1
```
Comment (JasonGross) [inline gbmi/exp_max_of_n/train.py:141]: Why this seed?  (Leave a comment about how (non)arbitrary this is)

**Review by JasonGross — COMMENTED** [review]
Initial review comments

File: gbmi/utils/__init__.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+from transformer_lens import HookedTransformer
+
+import numpy as np
+import torch
+from jaxtyping import Float
+from lightning import Callback
+from numpy.random import Generator
+from torch import Tensor
+
+PROJECT_ROOT = Path(__file__).parent.parent.parent
```
Comment (JasonGross) [inline gbmi/utils/__init__.py:20]: This seems fragile, especially since it goes outside the top-level package, and with respect to refactoring.  Maybe we want to have a `get_git_project_root`?  Or ensure it's a subdirectory of the package?

File: gbmi/utils/hash.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+"""
+import dataclasses
+import datetime
+import hashlib
+import json
+from collections.abc import Collection
+from typing import Any
+from typing import Dict
+
+# Implemented for https://github.com/lemon24/reader/issues/179
```
Comment (JasonGross) [inline gbmi/utils/hash.py:24]: Did you write this code @euanong ?  Do we need to stick some license notice at the top?

**Review by JasonGross — COMMENTED** [review]
Initial review comments

**Comment by JasonGross** [comment]
I'm going to merge this now so we can build on it, code review can be fixed in another PR

File: gbmi/__init__.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+
+
+
+
+
```
[the remark this answers is not in the document]
Comment (euanong) [reply gbmi/__init__.py:5]: fixed

File: gbmi/analysis_tools/decomp.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+def analyze_svd(
+    M,
+    descr="",
+    scale_by_singular_value=True,
+    colorscale="Picnic_r",
+    singular_color="blue",
+    renderer=None,
+):
+    U, S, Vh = torch.linalg.svd(M)
+    V = gbmi.utils.T
```
[the remark this answers is not in the document]
Comment (euanong) [reply gbmi/analysis_tools/decomp.py:19]: fixed

File: gbmi/analysis_tools/l1h1.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import torch
```
[the remark this answers is not in the document]
Comment (euanong) [reply gbmi/analysis_tools/l1h1.py:1]: fixed

File: gbmi/verification_tools/general.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+import torch
+from torchtyping import TensorType
```
[the remark this answers is not in the document]
Comment (euanong) [reply gbmi/verification_tools/general.py:2]: done
